### PR TITLE
Extract out data store metrics for re-usability

### PR DESCRIPTION
**Pull request by iaroslav-ciupin**
# TL;DR
Adjust DataStore to allow initialising it multiple times during a single application run.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: storage/copy_impl.go [hunk 1]
```diff
@@ -64,18 +64,11 @@ func (c copyImpl) CopyRaw(ctx context.Context, source, destination DataReference
...
+func newCopyMetrics(scope promutils.Scope) *copyMetrics {
+	return &copyMetrics{
 		CopyLatency:                  labeled.NewStopWatch("overall", "Overall copy latency", time.Millisecond, scope, labeled.EmitUnlabeledMetric),
 		ComputeLengthLatency:         labeled.NewStopWatch("length", "Latency involved in computing length of content before writing.", time.Millisecond, scope, labeled.EmitUnlabeledMetric),
 		WriteFailureUnrelatedToCache: scope.MustNewCounter("write_failure_unrelated_to_cache", "Raw store write failures that are not caused by ErrFailedToWriteCache"),
 		ReadFailureUnrelatedToCache:  scope.MustNewCounter("read_failure_unrelated_to_cache", "Raw store read failures that are not caused by ErrFailedToWriteCache"),
 	}
 }
-
-func newCopyImpl(store RawStore, metricsScope promutils.Scope) copyImpl {
```
Comment (katrogan) [inline storage/copy_impl.go:76]: can we keep this constructor but pass in the set of already initialized copyMetrics?

File: storage/protobuf_store.go [hunk 2]
```diff
@@ -81,17 +79,9 @@ func (s DefaultProtobufStore) WriteProtobuf(ctx context.Context, reference DataR
 	return nil
 }
 
-func NewDefaultProtobufStore(store RawStore, metricsScope promutils.Scope) DefaultProtobufStore {
+func NewDefaultProtobufStore(store RawStore, metrics *DataStoreMetrics) DefaultProtobufStore {
```
Comment (katrogan) [inline storage/protobuf_store.go:82]: should this take in `protoMetrics *protoMetrics` only?

File: storage/rawstores.go [hunk 3]
```diff
@@ -54,8 +56,51 @@ func createHTTPClient(cfg HTTPClientConfig) *http.Client {
...
+
+			WriteFailure: labeled.NewCounter("write_failure", "Indicates failure in storing/PUT for a given reference", scope, labeled.EmitUnlabeledMetric, failureTypeOption),
+			WriteLatency: labeled.NewStopWatch("write", "Time to write an object irrespective of size", time.Millisecond, scope, labeled.EmitUnlabeledMetric),
+		},
+	}
+}
+
 // NewDataStore creates a new Data Store with the supplied config.
-func NewDataStore(cfg *Config, metricsScope promutils.Scope) (s *DataStore, err error) {
+func NewDataStore(cfg *Config, metrics *DataStoreMetrics) (s *DataStore, err error) {
```
Comment (katrogan) [inline storage/rawstores.go:103]: let's keep the constructor as is, but also add a method `func (d *DataStore) RefreshConfig(cfg *Config) error` that can re-use code defined in this constructor (this shared code can be pulled out to a separate private func for re-use)

**Review by katrogan — COMMENTED** [review]
looks good, just a few comments

**Comment by kumare3** [comment]
Why do we need to do this? It seems it can already be reinitialized - pass a different scope

File: storage/rawstores.go [hunk 4]
```diff
@@ -54,27 +56,53 @@ func createHTTPClient(cfg HTTPClientConfig) *http.Client {
...
+	stowMetrics  *stowMetrics
+}
 
-		protoStore := NewDefaultProtobufStore(newCachedRawStore(cfg, rawStore, metricsScope), metricsScope)
-		return NewCompositeDataStore(NewURLPathConstructor(), protoStore), nil
+// newDataStoreMetrics initialises all metrics required for DataStore
+func newDataStoreMetrics(scope promutils.Scope) *dataStoreMetrics {
+	failureTypeOption := labeled.AdditionalLabelsOption{Labels: []string{FailureTypeLabel.String()}}
+	return &dataStoreMetrics{
+		cacheMetrics: &cacheMetrics{
```
Comment (katrogan) [inline storage/rawstores.go:70]: does it make sense to add metrics constructors, like `NewCacheMetrics` alongside the store constructors in their respective files so that the newDataStoreMetrics can just call those and the metrics initialization code is better co-located with their usage? similar to what we're currently doing with `newCopyMetrics`

**Review by katrogan — COMMENTED** [review]
looks good, just one suggestion

File: storage/rawstores.go [hunk 5]
```diff
@@ -84,3 +84,28 @@ func NewCompositeDataStore(refConstructor ReferenceConstructor, composedProtobuf
 		ComposedProtobufStore: composedProtobufStore,
 	}
 }
+
+func (ds *DataStore) RefreshConfig(cfg *Config) error {
```
Comment (katrogan) [inline storage/rawstores.go:88]: sorry one last request, @pmahindrakar-oss mentioned that this isn't thread-safe which is a great catch - do you mind adding a comment saying so?

**Review by katrogan — APPROVED** [review]
